### DISCUSSION TITLE
Spawn an entity by another entity.

### DISCUSSION THREAD

**Discussion by RyanAelen** [han-yaeger/yaeger#242] [Q&A]
I see that you can use an entity spawner to add new entities during the game.
However, with this one you have to use a time interval. Isn't there an easy way within an entity to add a new entity to the scene that has already started?

I have a composite entity, when it changes state I want it to remove some of its childs entity. When it changes state again, these child entities must be spawned again.

Shortened class

```
public abstract class BaseSpriteEntity extends CompositeEntity {
    protected SpriteEntity sprite;
    protected boolean blocksEntitys = true;
    protected HitBox hitbox;

    @Override
    protected void setupEntities() {
        addEntity(sprite);
        addHitbox();
    }

    private void addHitbox() {
        hitbox = new HitBox(getLocationInScene(), this);
        addEntity(hitbox);
    }

    public void setBlocks(boolean blocks) {
        this.blocksEntitys = blocks;
        if (hitbox != null && !blocks) hitbox.remove();
        if (hitbox == null && blocks) addHitbox();
    }
}
```
`public class HitBox extends CompositeEntity `
the `setBlocks()` only works when removing and not when re-inserting while the game is running

---

Ik zie dat je een entity spawner kan gebruiken om nieuwe entity's toe te kunnen voegen tijdens het spel.
Bij deze moet je echter een tijdsinterval gebruiken. Is er niet een makkelijke manier om binnen een entity een nieuwe entity toe te voegen aan de scène die al gestart is?

Ik heb een composite entity, wanneer deze van state verander wil ik dat die een aantal van zijn childs entity weghaalt. Wanneer die nog een keer van status veranderd moeten deze child entity opnieuw in spawned.

Ingekorte class

```
public abstract class BaseSpriteEntity extends CompositeEntity {
    protected SpriteEntity sprite;
    protected boolean blocksEntitys = true;
    protected HitBox hitbox;

    @Override
    protected void setupEntities() {
        addEntity(sprite);
        addHitbox();
    }

    private void addHitbox() {
        hitbox = new HitBox(getLocationInScene(), this);
        addEntity(hitbox);
    }

    public void setBlocks(boolean blocks) {
        this.blocksEntitys = blocks;
        if (hitbox != null && !blocks) hitbox.remove();
        if (hitbox == null && blocks) addHitbox();
    }
}
```
`public class HitBox extends CompositeEntity `
de `setBlocks()` werkt alleen bij het weg halen en niet bij het opnieuw plaatsen terwel het spel draaied

**Comment by meronbrouwer** [comment]
Right now, the way to spawn entities from another entity would be to call addEntity(YaegerEntity) on the scene. The EntitySpawners were explicitly designed for spawning them at a given interval. And since entities should be added to a scene, only a scene could implement the SpawnerContainer interface.

Right now, I'm thinking about a redesign in which we supply an EntitySpawner and a DynamicEntitySpawner. The first one exposes a method that should be called by the user to spawn an entity. The DynamicEntitySpawner shows the behaviour as is implemented right now. Moreover the SpawnerContainer will also be supported on an entity, so they no longer need a reference to the scene.

What are your views on this?

**Reply by meronbrouwer** [reply]
Since a question related to this discussion pops up every now and then, lets clarify how to currently add (or spawn) Entities from an Entity.

Right now spawning Entities means adding them at a regular time interval. Since only s Scene can contain an Entity, only a Scene can become an `EntitySpawnerContainer` and only a Scene exposes the `addEntity(YaegerEntity)` method.
With this in mind, if a Entity (lets call this one `EntityA`) needs to spawn (add) another Entity (`EntityB`), `EntityA` should receive a reference to the Scene it has been added to.

```java
public class EntityA extends DynamicSpriteEntity {

    private YaegerScene yaegerScene;

    protected EntityA(String resource, Coordinate2D initialLocation, final YaegerScene yaegerScene) {
        super("entities/zombie-spritemap.png", initialLocation, 2, 12);
        this.yaegerScene = yaegerScene;
    }
}
```

With the instancefield `yaegerScene`, `EntityA` can call `yaegerScene.addEntity(new EntityB(getAnchorLocation()))` whenever it wants. In this example it passes its own anchorLocation as a paramater to the constructor of `EntityB`, which ensures `EntityB` appears at the same location as `EntityA`.
The last thing to keep in mind is that the method `addEntity(YaegerEntity)` has acces modifier `protected`, meaning `EntityA` would not be allowed to call it. To resolve this overwrite the method `addEntity(YaegerEntity)` in your own Scene and change the access modifer to `public.`
